### ISSUE TITLE
LightGbmMulticlassTrainer does not train correctly

### ISSUE THREAD

**Issue by Shadow992** [dotnet/machinelearning#2263]
### System information

- **OS version/distro**:
Windows 10 64bit

- **.NET Version (eg., dotnet --info)**: 
4.7.2 .Net Framework
ML.Net v0.9
Compiling with x64 and Release/Debug (does not matter)

### Issue

- **What did you do?**
I tried to get a simple LightGbmMulticlassTrainer working. I had have a look at several tutorials as I am quite new to this library but not to Machine Learning in general. 

- **What happened?**
When I train the dataset with LightGbmMulticlassTrainer, it will always return an accuracy of 33,3%, because it always predicts "Setosa". However what is strange is, that in general my approach works. When I use 

mlContext.MulticlassClassification.Trainers.StochasticDualCoordinateAscent()

instead of my LightGBM trainer it gives me accuracy of 100% (which is not too suprising as my testset and trainingset is the same).

### Source code / logs

            TextLoader textLoader = mlContext.Data.CreateTextReader(new TextLoader.Arguments()
            {
                Separator = ",",
                HasHeader = false,
                Column = new[]
                {
                    new TextLoader.Column("SepalLength", DataKind.R4, 0),
                    new TextLoader.Column("SepalWidth", DataKind.R4, 1),
                    new TextLoader.Column("PetalLength", DataKind.R4, 2),
                    new TextLoader.Column("PetalWidth", DataKind.R4, 3),
                    new TextLoader.Column("Label", DataKind.Text, 4)
                }
            });

            IDataView dataView = textLoader.Read(_dataPath);

            string featuresColumnName = "Features";
            var pipeline =  mlContext.Transforms.Conversion.MapValueToKey("Label")
                .Append(mlContext.Transforms.Concatenate("Features", "SepalLength", "SepalWidth", "PetalLength", "PetalWidth"))
                .Append(new LightGbmMulticlassTrainer(mlContext, "Label","Features"))
                .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel"));

            // Train the model.
            var model = pipeline.Fit(dataView);

            // Compute quality metrics
            var metrics = mlContext.MulticlassClassification.Evaluate(model.Transform(dataView));
            Console.WriteLine(metrics.AccuracyMacro);

**Comment by Ivanidzo4ka** [comment]
Hello @Shadow992 
I've tried to reproduce your problem, but I get proper metrics and predictions.
My project is in attachment:
[ConsoleApp12.zip](https://github.com/dotnet/machinelearning/files/2814776/ConsoleApp12.zip)

![image](https://user-images.githubusercontent.com/1523833/52017532-7781bc80-249c-11e9-8678-0f47d8ffa62c.png)

**Comment by Ivanidzo4ka** [comment]
Can you try my solution and check it works as expected or confirm it also doesn't work.

**Comment by Ivanidzo4ka** [comment]
What do you mean by latest version? My project point to 0.9 which is latest stable, are you talking about daily builds?

![image](https://user-images.githubusercontent.com/1523833/52071186-5326ee00-2537-11e9-9f4e-593c32e6c681.png)

Here results from dependency walker for lib_lightgbm.dll.
vcomp14. msvcp14 and vcruntime14 can be installed by https://www.microsoft.com/en-ca/download/details.aspx?id=48145 .

**Comment by Shadow992** [comment]
@Ivanidzo4ka 
I use nuget packages to manage the dependencies. Therefore I add ML by using nuget and as ML has some dependencies I install them over nuget, too. 

If I install the version you provided in project it "runs" except the exception with "lib_lightgbm.dll not found". I was talking about updating the dependencies as they are not the latest:

![image](https://user-images.githubusercontent.com/16254551/52073137-0e2aa800-2587-11e9-991f-9136960e6dea.png)

When I update all of them and retry it gives me the behaviour originally mentioned.

I am also quite sure to have all needed Redists. As I am developing for long time C# and C++ projects on this machine. But I redownloaded and tried to install. However it said: "You already have installed this or a similar version"
Therefore this should work, too.

I just retried a few times and the "Missing DLL thing" now disappeared. it seems to be a sporadic error or something similar. However now your exact version gives me the same result:

![image](https://user-images.githubusercontent.com/16254551/52074029-403d0980-2589-11e9-9a23-33a81d42daf7.png)

**Comment by Ivanidzo4ka** [comment]
Sorry for confusion. Technically you don't have to update nugets to latest, And I personally prefer not to do that. If we have reference on A version 11.0, and there is A with version 12.0, there is no guarantee version 12 has method C() which is present in 11.0 version and which we invoke, or maybe it does, but it do completely opposite thing. Which usually not the case, but still.

Did by any chance you updated LightGBM nuget we reference as well? From 2.2.1.1 to higher version? Since that can be main reason why you have this issue. I think @singlis looked into updating to latest one, but something is stopping us from that step.

If you take my project, it will initially complain regarding not present nugets. You can solve it by right click on project -> Manage Nuget Packages, and that window should pop up you suggestion to restore projects. I've done it on clean machine and it worked fine for me.

**Comment by Ivanidzo4ka** [comment]
Even if I update all nugets to latest one, I still get good results. Can you zip your project and it bin folder and add it to discussion?

**Comment by Shadow992** [comment]
@Ivanidzo4ka 
I am currently debugging the code and it seems like the regressionTrees are doing strange things, have a look at this:

![image](https://user-images.githubusercontent.com/16254551/52075249-1df8bb00-258c-11e9-80cf-b18fa5e930ad.png)

The split threshold seems way too high. I suspect some "local problems whenr eading the CSV file". In more detail I suspect, that the TextReader interprets "." as "delimiter for thousands" (like this 1.000.000) and not as a floating point number. In germany we use similar things. I will keep you up to date when more debgugging brings more insights.

Sharing the project will probably not be helpful as the project you posted gives me now the same results as my original project did.

**Comment by Ivanidzo4ka** [comment]
our data loader should work fine, since it's appear only for LightGBM but works fine for SDCA. LightGBM is native library, and they have zero respect to locale, while our C# code parse with local locale. I'm pretty at some point we got trees from lightgbm as strings and we parse them, and we should put en-us locale for parsing methods.

I'll take a look on my side as well, it's not so hard to replace locale for me on my machine.
Thank you for reporting this and sticking to issue!

**Comment by Shadow992** [comment]
I do not know if it helps, but also the splitgain looks quite big:

![image](https://user-images.githubusercontent.com/16254551/52076931-1d622380-2590-11e9-8410-a49d2c0adbb8.png)

Whereby I found this "gain" as "InnerArgs" (to be honest I cannot remember where exactly I found it... These are way too much classes and I am way to unfamiliar with this library...):

`unbalanced_sets=False min_split_gain=0 max_depth=0 min_child_weight=0.1 subsample_freq=0 subsample=1 feature_fraction=1 reg_lambda=0.01 reg_alpha=0 scale_pos_weight=1 boosting_type=gbdt max_bin=255 verbose=-1 seed=2014381767 sigmoid=0.5 label_gain=0,3,7,15,31,63,127,255,511,1023,2047,4095 use_missing=False min_data_per_group=100 max_cat_threshold=32 cat_smooth=10 cat_l2=10 learning_rate=0.2 num_leaves=20 min_data_per_leaf=5 num_class=3 objective=multiclassova metric=multi_error`

Probably this does not mean too much as it looks like 2^x-1, but still I wanted to mention it, maybe it helps.
